File: leetcode/199.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> RightSideView(TreeNode root) {
        if (root == null) {
            return new List<int>();
        }
        List<int> rightSides = new List<int>();
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while (queue.Count > 0) {
            int count = queue.Count;
            for (int i = 0; i < count; i++) {
                TreeNode node = queue.Dequeue();
                if (node.left != null) {
                    queue.Enqueue(node.left);
                }
                if (node.right != null) {
                    queue.Enqueue(node.right);
                }
                if (i == count - 1) {
                    rightSides.Add(node.val);
                }
            }
        }
        return rightSides;
    }
}

File: leetcode/199.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int depth = 0;

    private void traverse(TreeNode node, int currentDepth, List<Integer> rightSides) {
        if (node == null) {
            return;
        }
        if (currentDepth > depth) {
            rightSides.add(node.val);
            depth++;
        }
        traverse(node.right, currentDepth + 1, rightSides);
        traverse(node.left, currentDepth + 1, rightSides);
    }

    public List<Integer> rightSideView(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<Integer> rightSides = new ArrayList<>();
        rightSides.add(root.val);
        traverse(root, 0, rightSides);
        return rightSides;
    }
}